text/3634-scoped-impl-trait-for-type.md

```diff
@@ -0,0 +1,3436 @@
+```
+
+where `fmt` receives the specified scoped implementation by observing it through the `T: Debug` bound on its implementing `impl` block.
+
+If no observing bound exists, code of this form should produce a warning spanning the `Trait in module` tokens. (see [unused-scoped-implementation])
+
+## Type aliases are opaque to scoped implementations
+[type-aliases-are-opaque-to-scoped-implementations]: #type-aliases-are-opaque-to-scoped-implementations
+
+As scoped `impl Trait for Type` is a fully lexically-scoped feature, the *implementation environment* present in a scope does not affect types hidden behind a type alias, except for the top-level type directly:
+
+```rust
+trait Trait {
+    fn method(&self) -> &str;
+}
+
+impl Trait for Type {
+    fn method(&self) -> &str {
+        "global"
+    }
+}
+
+mod m1 {
+    use super::Type;
+
+    pub type Alias = [Type; 1];
+}
+
+mod m2 {
+    use super::{Type, Trait};
+
+    pub use impl Trait for Type {
+        fn method(&self) -> &str {
+            "scoped"
+        }
+    }
+
+    pub use impl<T: Trait> Trait for [T; 1] {
+        fn method(&self) -> &str {
+            self[0].method()
+        }
+    }
+}
+
+fn main() {
+    use m1::Alias;
+    use m2::{
+        impl Trait for Type,
+        impl Trait for [Type; 1],
+    };
+
+    assert_eq!([Type].method(), "scoped");
+    assert_eq!(Alias::default().method(), "global");
+}
+```
+
+Scoped implementations may still be observed through bounded generic type parameters on the type alias itself. (see [binding-choice-by-implementations-bounds])
+
+## `TypeId` of generic type parameters' opaque types
+[typeid-of-generic-type-parameters-opaque-types]: #typeid-of-generic-type-parameters-opaque-types
+
+In addition to the type identity of the specified type, the `TypeId` of opaque generic type parameter types varies according to the captured *implementation environment*, but *only according to implementations that are relevant to their bounds (including implicit bounds)*, so that the following program runs without panic:
+
+```rust
+use std::any::TypeId;
+
+#[derive(Default)]
+struct Type;
+trait Trait {}
+impl Trait for Type {}
+
+#[derive(Default)]
+struct Generic<T>(T);
+
+mod nested {
+    pub(super) use impl super::Trait for super::Type {}
+}
+
+// `A` and `B` are distinct due to different captured implementation environments.
+type A = Generic<Type>;
+type B = Generic<Type as Trait in nested>;
+
+fn no_bound<T: 'static, U: 'static>(_: Generic<T>, _: Generic<U>) {
+    assert_eq!(TypeId::of::<T>(), TypeId::of::<U>());
+    assert_ne!(TypeId::of::<Generic<T>>(), TypeId::of::<Generic<U>>());
+
+    assert_eq!(TypeId::of::<T>(), TypeId::of::<Type>());
+    assert_eq!(TypeId::of::<U>(), TypeId::of::<Type>());
+}
+
+fn yes_bound<T: Trait + 'static, U: Trait + 'static>(_: Generic<T>, _: Generic<U>) {
+    assert_ne!(TypeId::of::<T>(), TypeId::of::<U>());
+    assert_ne!(TypeId::of::<Generic<T>>(), TypeId::of::<Generic<U>>());
```

Comment:
```suggestion
    assert_ne!(TypeId::of::<Generic<T>>(), TypeId::of::<Generic<U>>()); // ⚠⚠
```

(The code block is messing with the suggestion, so I have to split this to do the edit here.)